internal/credit/postgres_connector/grant.go

```diff
@@ -0,0 +1,401 @@
+package postgres_connector
+
+import (
+	"context"
+	"fmt"
+	"math"
+
+	"entgo.io/ent/dialect/sql"
+
+	connector "github.com/openmeterio/openmeter/internal/credit"
+	"github.com/openmeterio/openmeter/internal/credit/postgres_connector/ent/db"
+	db_credit "github.com/openmeterio/openmeter/internal/credit/postgres_connector/ent/db/creditentry"
+	credit_model "github.com/openmeterio/openmeter/pkg/credit"
+)
+
+func (c *PostgresConnector) CreateGrant(ctx context.Context, namespace string, grant credit_model.Grant) (credit_model.Grant, error) {
+	// Lock ledger for the subject
+	logger := c.logger.With("operation", "createGrant", "namespace", namespace, "subject", grant.Subject, "id", grant.ID)
+	lock, err := c.lockManager.Obtain(ctx, namespace, grant.Subject)
+	if err != nil {
+		return credit_model.Grant{}, err
+	}
+	defer func() {
+		err = c.lockManager.Release(ctx, lock)
+		if err != nil {
+			logger.Error("failed to release lock", "error", err)
+		}
+	}()
+
+	q := c.db.CreditEntry.Create().
+		SetNamespace(namespace).
+		SetNillableID(grant.ID).
+		SetSubject(grant.Subject).
+		SetEntryType(credit_model.EntryTypeGrant).
+		SetType(grant.Type).
+		SetNillableParentID(grant.ParentID).
+		SetNillableProductID(grant.ProductID).
+		SetAmount(grant.Amount).
+		SetPriority(grant.Priority).
+		SetEffectiveAt(grant.EffectiveAt).
+		SetExpirationPeriodDuration(grant.Expiration.Duration).
+		SetExpirationPeriodCount(grant.Expiration.Count).
+		SetMetadata(grant.Metadata)
+	if grant.Rollover != nil {
+		q = q.SetRolloverType(grant.Rollover.Type).
+			SetNillableRolloverMaxAmount(grant.Rollover.MaxAmount)
+	}
+	entity, err := q.Save(ctx)
+	if err != nil {
+		return grant, fmt.Errorf("failed to create grant: %w", err)
+	}
+
+	grant, err = mapGrantEntity(entity)
+	if err != nil {
+		return grant, fmt.Errorf("failed to map grant entity: %w", err)
+	}
+	return grant, nil
+}
+
+// TODO: use grant ID as an argument to void grant
+func (c *PostgresConnector) VoidGrant(ctx context.Context, namespace string, grant credit_model.Grant) (credit_model.Grant, error) {
```

Comment:
User Story: I think it is a question how we want to handle negative balances.

If a subject has an X amounts of grant from a single source and Y amounts of usage (Y>0) and the operator voids a grant that means that the subject will go into a negative balance.

Given that these are prepaid accounts making a customer pay for the negative balance can be troublesome. We can say that we allow for negative balances, but is this something our users would want or do they want an error in such cases?